mesonbuild/utils/universal.py

```diff
@@ -566,29 +564,23 @@ def miss_defaulting(self) -> "PerThreeMachineDefaultable[T.Optional[_T]]":
     def matches_build_machine(self, machine: MachineChoice) -> bool:
         return self.build == self[machine]
 
-    def __repr__(self) -> str:
-        return f'PerThreeMachine({self.build!r}, {self.host!r}, {self.target!r})'
-
 
+@dataclasses.dataclass(eq=False, order=False)
 class PerMachineDefaultable(PerMachine[T.Optional[_T]]):
     """Extends `PerMachine` with the ability to default from `None`s.
     """
-    def __init__(self, build: T.Optional[_T] = None, host: T.Optional[_T] = None) -> None:
-        super().__init__(build, host)
 
-    def default_missing(self) -> "PerMachine[_T]":
+    build: T.Optional[_T] = None
+    host: T.Optional[_T] = None
+
+    def default_missing(self) -> PerMachine[_T]:
         """Default host to build
 
         This allows just specifying nothing in the native case, and just host in the
         cross non-compiler case.
         """
-        freeze = PerMachine(self.build, self.host)
-        if freeze.host is None:
-            freeze.host = freeze.build
-        return freeze
-
-    def __repr__(self) -> str:
-        return f'PerMachineDefaultable({self.build!r}, {self.host!r})'
+        assert self.build is not None, 'Cannot fill in missing when all fields are empty'
+        return PerMachine(self.build, self.host if self.host is not None else self.build)
```

Comment:
Personally, I find it less verbose to write it like:

```suggestion
        return PerMachine(self.build, self.host or self.build)
```

Reply:
That will do the wrong thing if the host is a falsy but non-null value, like `PerMachineDefaultable([], []).default_missing()`, where `host` will become an alias of `build` instead of being it's own value.

Reply:
I see. I didn't realized it was a generic class...